controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/CommitRequestHandler.java

```diff
@@ -0,0 +1,349 @@
+                            future.completeExceptionally(cause);
+                        }
+                    } else {
+                        log.debug("Successfully committed transactions on epoch {} on stream {}/{}", epoch, scope, stream);
+                        if (processedEvents != null) {
+                            processedEvents.offer(event);
+                        }
+                        future.complete(r);
+                    }
+                });
+
+        return future;
+    }
+
+    private CompletableFuture<Void> tryCommitTransactions(final String scope,
+                                                          final String stream,
+                                                          final int txnEpoch,
+                                                          final OperationContext context) {
+        // try creating txn commit list first. if node already exists and doesnt match the processing in the event, throw operation not allowed.
+        // This will result in event being posted back in the stream and retried later. Generally if a transaction commit starts, it will come to
+        // an end.. but during failover, once we have created the node, we are guaranteed that it will be only that transaction that will be getting
+        // committed at that time.
+        return streamMetadataStore.getState(scope, stream, true, context, executor)
+                .thenCompose(state -> {
+                    CompletableFuture<List<UUID>> txnListFuture = createRecordAndGetCommitTxnList(scope, stream, txnEpoch, context);
+
+                    CompletableFuture<Void> commitFuture = txnListFuture
+                            .thenCompose(txnList -> {
+                                if (txnList == null) {
+                                    // reset state conditionally in case we were left with stale committing state from a previous execution
+                                    // that died just before updating the state back to ACTIVE but after having completed all the work.
+                                    return streamMetadataStore.resetStateConditionally(scope, stream, State.COMMITTING_TXN, context, executor);
+                                } else {
+                                    // Once state is set to committing, we are guaranteed that this will be the only processing that can happen on the stream
+                                    // and we can proceed with committing outstanding transactions collected in the txnList step.
+                                    CompletableFuture<Void> future;
+                                    // if state is sealing, we should continue with commit so that we allow for completion of transactions
+                                    // in commit state.
+                                    if (state.equals(State.SEALING)) {
```

Comment:
Is it possible that we have txns in the `COMMITTING` state, even though the stream has been sealed? That would be a bug, I'd say.

Reply:
No, seal task waits for all transactions on the stream to finish (either commit or abort) so those in COMMITTING will be committed and those in ABORTING will be aborted.